### PR TITLE
Sheets

### PR DESCRIPTION
**Pull request by laurents17**
In need of some feedback and assistance with how to arrange the code in my placements_import model so I can upload a sheet full of placements without specifying a licensee for each one.

### CODE REVIEW

**Comment by laurents17** [comment]
@jrmcgarvey @micronix @leila-alderman

**Comment by laurents17** [comment]
@jrmcgarvey @leila-alderman @micronix 
Requesting a code review for help with the import of a spreadsheet full of licensees using roo gem. The import and export functions work correctly for the placements table, and the code for importing licensees is the same. I made an attempt to add strong params to the licensees import and it still responds the same way (see next paragraph). I also added the admin authorization statement to both placements_imports_controller and licensees_imports_controller. 

When a licensees_import fails, it should render :new and show the list of errors in that view (as outlined in the controller, and the save method in the model). Instead, it just follows the licensees_path and goes back to the index view without saving the import. Any idea why this would happen, when it functions correctly for the placements_import and the code is essentially the same? 

In the licensees_import logs (attached) you can see that the file is entering params but then it is redirecting to the index view. (Maybe that means it is a routing issue? But the links are set up the same way for placements_import and licensees_import). In contrast, the placements_import logs show that after that line, the records are being stored correctly in the placements table.

licensees_import logs
placements_import logs

File: app/controllers/licensees_controller.rb [hunk 1]
```diff
@@ -1,56 +1,66 @@
 # frozen_string_literal: true
 
 class LicenseesController < ApplicationController
+
   before_action :authenticate_user!
   before_action :authorize_admin, only: %i[new edit]
+    
+    def new
+        @licensee = Licensee.new
+    end
```
Comment (leila-alderman) [inline app/controllers/licensees_controller.rb]: [Style]
The indentation is off here. Each indentation should be 2 spaces, not four. You can change the default for this on your text editor.

File: app/controllers/licensees_controller.rb [hunk 2]
```diff
@@ -1,56 +1,66 @@
...
+    def create
+        Licensee.create(
+            facility_name: params[:licensee][:facility_name],
+            contact_name: params[:licensee][:contact_name],
+            address: params[:licensee][:address],
+            city: params[:licensee][:city],
+            state: params[:licensee][:state],
+            zip: params[:licensee][:zip],
+            phone: params[:licensee][:phone],
+            fax: params[:licensee][:fax],
```
Comment (leila-alderman) [inline app/controllers/licensees_controller.rb]: Please use strong parameters. 

Using strong parameters is important for form input security and keeping your controller code DRY.
To start, you can read through the short section on strong parameters from the Rails Guides: https://edgeguides.rubyonrails.org/action_controller_overview.html#strong-parameters
Next, to see them in action, read through this section from Hartl's Rails Tutorial book: https://www.railstutorial.org/book/sign_up#sec-strong_parameters

File: app/controllers/licensees_controller.rb [hunk 3]
```diff
@@ -1,56 +1,66 @@
...
+        @licensee.facility_name = params[:licensee][:facility_name]
+        @licensee.contact_name = params[:licensee][:contact_name]
+        @licensee.address = params[:licensee][:address]
+        @licensee.city = params[:licensee][:city]
+        @licensee.state = params[:licensee][:state]
+        @licensee.zip = params[:licensee][:zip]
+        @licensee.phone = params[:licensee][:phone]
+        @licensee.fax = params[:licensee][:fax]
+        @licensee.save
+        redirect_to licensee_path(params[:id])
```
Comment (leila-alderman) [inline app/controllers/licensees_controller.rb]:
```suggestion
        redirect_to licensee_path(@licensee)
```

File: app/controllers/licensees_imports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def create
+    @licensees_import = LicenseesImport.new(licensees_import_params)
+    #@licensees_import = LicenseesImport.new(params[:licensees_import])
+    #byebug
+    if @licensees_import.save
+      redirect_to licensees_path
+    else
+      render :new
```
Comment (leila-alderman) [inline app/controllers/licensees_imports_controller.rb]: Consider adding a flash message here (and in the view) to let the user know what went wrong.

https://www.railstutorial.org/book/sign_up#sec-the_flash

File: app/controllers/placements_imports_controller.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  def new
+    @placements_import = PlacementsImport.new
+  end
+
+  def create
+    @placements_import = PlacementsImport.new(params[:placements_import])
+    if @placements_import.save
+      redirect_to placements_path
+    else
+      render :new
```
Comment (leila-alderman) [inline app/controllers/placements_imports_controller.rb]: Consider adding a flash here.

File: app/views/licensees_imports/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+<p>
+Note: Only import spreadsheets that were previously downloaded from this
+  application.
+</p>
+
+<h3>Requirements</h3>
+<ul>
+  <li>Here you could list the validations you have in place for your model</li>
+  <li>That way your user has a better chance of successful import</li>
```
Comment (leila-alderman) [inline app/views/licensees_imports/new.html.erb]: This text should be updated from the placeholder.

File: app/views/placements_imports/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+<p>
+Note: Only import spreadsheets that were previously downloaded from this
+  application.
+</p>
+
+<h3>Requirements</h3>
+<ul>
+  <li>Here you could list the validations you have in place for your model</li>
+  <li>That way your user has a better chance of successful import</li>
```
Comment (leila-alderman) [inline app/views/placements_imports/new.html.erb]: This text should be updated from the placeholder.

File: config/routes.rb [hunk 8]
```diff
@@ -13,6 +15,9 @@
   # end
   resources :placements, :comments, :licensees
   resources :placements_imports, only: [:new, :create]
+  resources :licensees_imports, only: [:new, :create]
```
Comment (leila-alderman) [inline config/routes.rb]: This seems redundant with lines 2 and 3 above. 

Please delete lines 2-5 and keep the resource versions here.

File: config/routes.rb [hunk 9]
```diff
@@ -13,6 +15,9 @@
...
+  root to: 'placements#index'
```
Comment (leila-alderman) [inline config/routes.rb]: The `root` line should always go at the top of the `routes` file.

**Pull request merged**